### DISCUSSION TITLE
.NET Standard 2.0支持

### DISCUSSION THREAD

**Discussion by MiharaIchiro** [Hoshikawa-Kaguya/Sora#37] [Q&A]
現時在我使用的環境下無法以<kbd>.net5</kbd>執行，可否對Sora額外添加<kbd>.NET Standard 2.0</kbd>的支持？

**Comment by Yukari316** [comment]
应该不会对 `.NET Standard 2.0` 进行支持，因为 `.Net5` 与 `.NET Standard 2.0` 的API和开发方式都不同，并且框架也使用了大量的 `C#9` 语法

详细的原因可以参考这篇文章：https://docs.microsoft.com/zh-cn/dotnet/standard/net-standard#net-5-and-net-standard

并且我想问一下你是在什么环境下运行 `.Net5` 程序的

**Reply by MiharaIchiro** [reply]
> 并且我想问一下你是在什么环境下运行 `.Net5` 程序的

``Windows10``<kbd>21H1</kbd> COM元件形式依附在``explorer.exe``下運行的工具列，但受限於我的技術，不能完全指出成因或有任何缺漏失誤而導致其以``.Net5`` ``net5-windows``無法正常加載。

具體地說，目前主程序是``net462``，所以無法引入``.Net5``的Sora框架，若主程序更改作``.Net5``則影子都沒有。我覺得主要成因是``explorer.exe`` ``Loadlibrary``的時候沒能成功讀取``.Net``.``WindowsDesktop``架構所必須的 ``runtimeconfig.json`` 引致。

出於技術限制，目前情況下主程序架構無法更變至``.Net5``。及又若主程序使用``.NET Standard 2.0``的``Package``是可以正常運行的，故而希望Sora對``.NET Standard 2.0``進行支持。

**Reply by MiharaIchiro** [reply]
關於``C#9`` 依賴 ``.Net5``的問題，網絡有一些[解決方案](https://blog.ndepend.com/using-c9-record-and-init-property-in-your-net-framework-4-x-net-standard-and-net-core-projects/)  。經fork Sora進行相應變更後，``C#9`` 引伸的大部分問題有效處理，除了``Deconstruction``。以及 ``.Net5`` 的 ``HashCode`` 、 及 ``ValueTask.FromResult`` 仍需修改以支持``.NET Standard 2.0``。

**Comment by xh321** [comment]
你如果需要生成单个文件，可以试试使用“发布”功能，将主工程切换至.Net 5 并按下图配置编译选项
![DOHTM 7Y`C~ZI{~%BCUQ~ 1](https://user-images.githubusercontent.com/37585529/130958602-e69f28c1-ac94-424c-92d5-110732222b56.png)
这样编译出来是没有额外的json文件的。
如果是因为COM可见性而导致的问题，可以参考下面这个项目
https://github.com/kyourek/ComVisibleNetStandard
来自这个回答：https://stackoverflow.com/questions/44861236/make-a-net-standard-library-com-visible

**Reply by xh321** [reply]
注：目标运行时要选择windows

**Reply by MiharaIchiro** [reply]
感謝你提供的方法，因為我的項目引用到``WindowsForms``以及``WPF``的關系，``TargetFramework``需要設定為``net5-windows``。
而在``net5-windows``下沒有``Self-contained``參數。而COM可見性則沒有問題。
